keps/sig-network/1669-proxy-terminating-endpoints/README.md

```diff
@@ -0,0 +1,378 @@
+
+### Non-Goals
+
+* Changing the behavior of how pods terminate.
+* Handling terminating endpoints for other consumers of the EndpointSlice API, such as ingress controllers or external load balancers.
+
+## Proposal
+
+This KEP proposes that if all endpoints for a given Service scoped to its traffic policy are terminating (i.e. pod.DeletionTimestamp != nil), then all traffic should be sent to
+terminating Pods that are still Ready. Note that the EndpointSlice API introduced a new condition called "Serving" which is semantically equivalent to "Ready" except that the Ready condition
+must always be "False" for terminating pods for compatibility reasons. For consumers of the EndpointSLice API that want to route traffic strictly based on a Pod's readiness ignoring
+it's terminating state, they should be reading the Serving condition going forward. Below are some examples to help illustrate the proposed behavior:
+
+### Example: only some endpoints terminating when traffic policy is "Cluster"
+
+When the traffic policy is "Cluster" and some endpoints are terminating, all traffic should be routed to the ready endpoints that are not terminating..
+
+### Example: only some endpoints terminating on a node when traffic policy is "Local"
+
+When the traffic policy is "Local" and some endpoints are terminating within a single node, traffic should be routed to ready endpoints on that node that are not terminating.
+
+### Example: all endpoints terminating and traffic policy is "Cluster"
+
+When the traffic policy is "Cluster" and all endpoints are terminating, then traffic should be routed to any terminating endpoint that is ready.
+
+### Example: all endpoints terminating on a node when traffic policy is "Local"
+
+When the traffic policy is "Local" and all endpoints are terminating within a single node, then traffic should be routed to any terminating endpoint that is ready on that node.
+
+
+### Handling terminating endpoints that are not ready.
+
+It is worth noting that traffic should not be routed to terminating pods if their readiness probe is failing, even if it is the only endpoints remaining. This is to give workloads
+the flexibility/control to opt out of this behavior by either exiting immediately or failing the readiness probe when receiving SIGTERM from kubelet. This would also be counter-intuitive
+to the current understanding of readiness probes.
+
+
+### User Stories (optional)
+
+#### Story 1
+
+As a user I would like to do a rolling update of a Deployment fronted by a Service Type=LoadBalancer with ExternalTrafficPolicy=Local.
+If a node that has only 1 pod of said deployment goes into the `Terminating` state, all traffic to that node is dropped until either a new pod
+comes up or my cloud provider removes the node from the loadbalancer's node pool. Ideally the terminating pod should gracefully handle traffic to this node
+until either one of the conditions are satisfied.
```

Comment:
> It makes it less likely that there will be an interruption, but people don't want just "less likely".

Can you expand on this a bit? Nothing is guarenteed but what are the gaps right now that make it not "less likely" enough?  

> I think if we want to be able to do Local with no disruption, we need a feature that integrates more directly with the load balancer health checking somehow and actually ensures that the endpoint remains active until after the load balancer forgets about it.

The current feature as-is does interact with load balancer health checking. In that it fails the load balancer health checking if all endpoints are terminating, but it still routes to traffic to terminating ones in case the LB's health check is delayed:

https://github.com/kubernetes/kubernetes/blob/03fc2eeda270c9244dc23c75424db3a20bf61d7b/pkg/proxy/endpoints.go#L396

Reply:
> This _should_ reduce traffic loss pretty significanitly assuming SIGTERM is handled by the app

assuming the app handles SIGTERM, _and_ assuming that the pod's `terminationGracePeriodSeconds` is larger than `EndpointsController lag` + `kube-proxy processing Endpoints updates lag` + `load balancer health check interval`, which the pod has no way of ensuring.

I dunno. Maybe 30s is pretty much guaranteed to be enough... I had been thinking we tested this in OpenShift and found that it didn't always work, but maybe we just didn't have fully-working ProxyTerminatingEndpoints support at that point...

Reply:
Back when the alpha implementation of this feature was being developed, there was some testing done on OpenShift, see https://github.com/kubernetes/kubernetes/pull/97238#issuecomment-793897805 -- although that comment is pretty old so maybe things have changed. But my impression was that the feature was addressing the issue. If you're seeing otherwise on OpenShift, I'd love to dig deeper and figure out why. I would also think 30s is plenty